Add a HUK sample (#4912)
nrfconnect/sdk-nrf: samples/keys/hw_unique_key/README.rst

      #. If not using TF-M, observe the following output:

         .. code-block:: console

            Writing random keys to KMU
            Success!

      #. Observe the following output (the values for IV, key, and ciphertext will vary randomly):

         .. code-block:: console

            Generating random IV
            IV:
            ab8e7c595d6de7d297a00b6c

            Deriving key
            Key:
            8d6e8ad32f5dffc10df1de38a2556ba0e01cf4ed56ac1294b9c57965cddc519a
            Key ID: 0x7fffffe0

            Encrypting
            Plaintext:
            "Lorem ipsum dolor sit amet. This will be encrypted."
            4c6f72656d20697073756d20646f6c6f722073697420616d65742e2054686973
            2077696c6c20626520656e637279707465642e
            Ciphertext (with authentication tag):
            ea696bf71e106f7c74adfc3296556f4f25ac2c999e453e28c52fb41085ef7b89
            3cbadee1a505cf3ce1901f4bc2fcca4fb86ec68e4b5f1344bb66ef5ce733f47a
            33788a

         If an error occurs, the sample will print a message and raise a kernel panic.

Dependencies
*************

This sample uses the following libraries:

* :ref:`lib_hw_unique_key`
* :ref:`nrf_security`

> * :ref:`ug_tfm` when using a nonsecure board.

Reviewer: Two comments regarding this line
nonsecure -> non-secure
And then the reference... When the reference resolves, it says: Running applications with Trusted Firmware-M when using a nonsecure board.
I would suggest changing it to:
:ref:`Trusted Firmware-M <ug_tfm>`, when using a non-secure board.

Author: Thanks, done

> * :ref:Trusted Firmware-M <ug_tfm>, when using a non-secure board.

Reviewer: ```suggestion
* :ref:`Trusted Firmware-M <ug_tfm>`, when using a non-secure board.
```
It seems like the comment field does some conversions when I am using ``. Nice to know!

.. _hw_unique_key_usage:

Hardware Unique Key
###################

.. contents::
   :local:
   :depth: 2

This sample shows how to use a hardware unique keys (HUKs) to derive a key and use it for encryption, through the psa_crypto APIs.

Overview
********

> The sample does the following:
> 
> * Initialize the HW if applicable
>   ** When TFM is enabled, it does this automatically on boot
> * Generate HUKs if applicable
>   ** When TFM is enabled, it does this automatically on boot
> * Generate a random IV
> * Derive a key
>   ** When TFM is enabled, this is done using the psa_crypto APIs
>   ** When TFM is not enabled, native APIs are used, and the key is imported into psa_crypto.
> * Encrypt a sample string using the key_id received from psa_crypto.

Reviewer: ```suggestion
The sample derives a key used to encrypt a string as follows:

1. Initializing the hardware, if applicable.

   When TF-M is enabled, the hardware is initialized automatically when booting.

#. Generating HUKs, if applicable.

   When TF-M is enabled, the HUKs are generated automatically when booting.
   
#. Generating a random initialization vector (IV).

#. Deriving a key used for encryption.

   When TF-M is enabled, the key is derived using the psa_crypto APIs.
   Otherwise, the native APIs are used, and the key is imported into the psa_crypto.

#. Encrypting a sample string using the key_id received from the psa_crypto.
```

Author: ```suggestion
The sample goes through the following steps to derive a key and use it to encrypt a string:

1. Initializing the hardware, if applicable.

   When TF-M is enabled, the hardware is initialized automatically when booting.
   
#. Generating HUKs, if applicable.

   When TF-M is enabled, the HUKs are generated automatically when booting.
   
#. Generating a random initialization vector (IV).

#. Deriving a key used for encryption.

   When TF-M is enabled, the key is derived using the psa_crypto APIs.
   Otherwise, the native APIs are used, and the key is imported into psa_crypto.
   
#. Encrypting a sample string using the key_id received from the psa_crypto.
```

Requirements
************

The sample supports the following development kits:

.. table-from-rows:: /includes/sample_board_rows.txt
   :header: heading

>    :rows: nrf5340dk_nrf5340_cpuapp, nrf5340dk_nrf5340_cpuappns, nrf9160dk_nrf9160, nrf9160dk_nrf9160ns, nrf52840dk_nrf52840

Reviewer: A few other samples are adding support for nRF21540. Should it be done here as well?

Reviewer 2: No, this sample only works for a limited set of boards, and `nRF21540` is not one of them.

> Building and Running

Reviewer: ```suggestion
Building and running
```

********************

.. |sample path| replace:: :file:`samples/keys/hw_unique_key`

.. include:: /includes/build_and_run.txt

Testing
=======

|test_sample|

      1. |connect_terminal_specific|
      #. Reset the kit.

>       #. Observe that the following output (IV, key, and ciphertext will vary):

Reviewer: ```suggestion
      #. Observe that the following output (IV, key, and ciphertext will vary):
```
Observe the following output?

Author: The output will not be exactly this. The values for IV, key, and ciphertext will essentially be random for each time the kit is flashed.

Reviewer: No, I meant: Observe THAT the following output -> Observe the following output
No?

Author: Ah, thanks, fixed.

         .. code-block:: console

            Generating random IV
            IV:
            ab8e7c595d6de7d297a00b6c

            Deriving key
            Key:
            8d6e8ad32f5dffc10df1de38a2556ba0e01cf4ed56ac1294b9c57965cddc519a
            Key ID: 0x7fffffe0

            Encrypting
            Plaintext:
            "Lorem ipsum dolor sit amet. This will be encrypted."
            4c6f72656d20697073756d20646f6c6f722073697420616d65742e2054686973
            2077696c6c20626520656e637279707465642e
            Ciphertext (with authentication tag):
            ea696bf71e106f7c74adfc3296556f4f25ac2c999e453e28c52fb41085ef7b89
            3cbadee1a505cf3ce1901f4bc2fcca4fb86ec68e4b5f1344bb66ef5ce733f47a
            33788a

>          If an error occurs, the sample will print a message and it will raise a kernel panic.

Reviewer: ```suggestion
         If an error occurs, the sample will print a message and raise a kernel panic.
```

Dependencies
*************

This sample uses the following libraries:

* :ref:`lib_hw_unique_key`
* :ref:`nrf_security`

> * :ref:`ug_tfm`

Author: No it doesn't always use TFM. How should we mention that? Any ideas @mia-ko?

TF-M is used when using a non-secure board.

> * :ref:`ug_tfm`

Reviewer 2: Just a nit, having this here is not a blocker from my side

> * :ref:`ug_tfm`

Reviewer: Hm... But why not just say that?
Like in a note:
.. note::
  This sample uses the TF-M library only if a non-secure board is used.

> * :ref:`ug_tfm`

Reviewer: And not sure if we need to elaborate more on what a non-secure board is, of if this is clear in this case.

> * :ref:`ug_tfm`

Reviewer: Just re-opened this one to make it visible. Are we adding the note, or is it ok leaving it out?

> * :ref:`ug_tfm`

Author: Thanks, added a note inline

.. _hw_unique_key_usage:

Hardware unique key
###################

.. contents::
   :local:
   :depth: 2

This sample shows how to use a hardware unique key (HUK) to derive an encryption key through psa_crypto APIs.

Overview
********

The sample goes through the following steps to derive a key and use it to encrypt a string:

1. Initializing the hardware, if applicable.

   When TF-M is enabled, the hardware is initialized automatically when booting.

#. Generating HUKs, if applicable.

>    When TF-M is enabled, the HUKs are generated automatically when booting.

Reviewer 2: Should we explain that you need to reboot to get the key to cryptocell if `CONFIG_SECURE_BOOT` is set instead of TF-M ?

Author: Added some text about this

#. Generating a random initialization vector (IV).

#. Deriving a key used for encryption.

   When TF-M is enabled, the key is derived using the psa_crypto APIs.
   Otherwise, the native APIs are used, and the key is imported into psa_crypto.

#. Encrypting a sample string using the key_id received from the psa_crypto.

Requirements
************

The sample supports the following development kits:

.. table-from-rows:: /includes/sample_board_rows.txt
   :header: heading

>    :rows: nrf5340dk_nrf5340_cpuapp, nrf5340dk_nrf5340_cpuappns, nrf9160dk_nrf9160, nrf9160dk_nrf9160ns, nrf52840dk_nrf52840

Reviewer: What is nrf5340dk_nrf5340_cpuappns? Is it a typo?
I know we have nrf5340dk_nrf5340_cpuapp and nrf5340dk_nrf5340_cpunet

Reviewer 2: It is the non-secure variant of the board, and not a typo.

Reviewer: Hm... It's crashing the doc build. Let me check this. It must be a typo.

Reviewer: It should be: nrf5340dk_nrf5340_cpuapp_ns

Reviewer: Or, I've seen a couple of these around as well: nrf5340dk_nrf5340_cpuapp_and_cpuapp_ns

Author: Fixed, ns -> _ns

Building and running
********************

.. |sample path| replace:: :file:`samples/keys/hw_unique_key`

.. include:: /includes/build_and_run.txt

Testing
=======

|test_sample|

      1. |connect_terminal_specific|
      #. Reset the kit.
      #. If not using TF-M, observe the following output:

         .. code-block:: console

            Writing random keys to KMU
            Success!

>       #. Observe the following output (The values for IV, key, and ciphertext will vary randomly):

Reviewer: The -> the
